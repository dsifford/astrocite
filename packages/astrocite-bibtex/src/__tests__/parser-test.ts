// tslint:disable no-console
import * as fs from 'fs';
import { basename } from 'path';
import parse from '../parser';

interface TestCases {
    [filename: string]: string;
}

describe('BibTeX Parser', () => {
    let cases: TestCases = {};
    beforeAll(() => {
        const casenames = fs.readdirSync(`${__dirname}/cases`);
        for (const name of casenames) {
            cases = {
                ...cases,
                [basename(name, '.bib')]: fs.readFileSync(
                    `${__dirname}/cases/${name}`,
                    {
                        encoding: 'utf-8',
                    },
                ),
            };
        }
    });
    it('should parse a simple .bib file', () => {
        expect(parse(cases.simple)).toMatchSnapshot();
    });
    it('should parse a file without leading or trailing whitespace', () => {
        expect(parse(cases.no_leading_trailing_whitespace)).toMatchSnapshot();
    });
    it('should parse a valid date regardless of the order of date parts or what is available', () => {
        expect(parse(cases.dates)).toMatchSnapshot();
    });
    it('should parse and skip over preamble statements correctly in varying formats', () => {
        expect(parse(cases.preamble)).toMatchSnapshot();
    });
    it('should parse user-defined strings', () => {
        expect(parse(cases.strings)).toMatchSnapshot();
    });
    it('should default to "article" for unknown types', () => {
        expect(parse(cases.unknown_types)).toMatchSnapshot();
    });
    it('should parse known commands and skip unknown commands', () => {
        expect(parse(cases.commands)).toMatchSnapshot();
    });
    it('should parse dicratical commands', () => {
        expect(parse(cases.dicraticals)).toMatchSnapshot();
    });
    it('should parse names in various formats', () => {
        expect(parse(cases.names)).toMatchSnapshot();
    });
    it('should parse minified files', () => {
        expect(parse(cases.minified)).toMatchSnapshot();
    });
    it('should parse files with zero entries', () => {
        expect(parse(cases.no_entries)).toMatchSnapshot();
    });
    it('should parse empty files', () => {
        expect(parse(cases.empty)).toMatchSnapshot();
    });
    it('should parse entries without IDs', () => {
        expect(parse(cases.no_id)).toMatchSnapshot();
    });
    it('should parse files with CRLF line endings', () => {
        expect(parse(cases.crlf)).toMatchSnapshot();
    });
    it('should parse jabref files', () => {
        expect(parse(cases.jabref)).toMatchSnapshot();
    });
    it('should parse files with verbatim tildes, comments, odd characters in keys, super/subscript, no-dot i/j, and math-mode switches', () => {
        expect(
            parse(
                cases.tilde_comments_underscores_diacritics_sup_sub_nodoti_math_verbatim,
            ),
        ).toMatchSnapshot();
    });
});
