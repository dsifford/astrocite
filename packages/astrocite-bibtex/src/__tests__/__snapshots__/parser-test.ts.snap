// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BibTeX Parser should default to "article" for unknown types 1`] = `
Array [
  Object {
    "id": "should_default_to_article",
    "title": "My awesome type",
    "type": "article",
  },
  Object {
    "id": "should_also_be_article",
    "title": " Foobar type ",
    "type": "article",
  },
]
`;

exports[`BibTeX Parser should parse a file without leading or trailing whitespace 1`] = `
Array [
  Object {
    "id": "my_ref_id",
    "note": "Note in quotes",
    "number": "6",
    "title": "Title in braces",
    "type": "book",
  },
]
`;

exports[`BibTeX Parser should parse a simple .bib file 1`] = `
Array [
  Object {
    "DOI": "10.1097/CCM.0000000000000981",
    "URL": "http://content.wkhealth.com/linkback/openurl?sid=WKPTLP:landingpage&an=00003246-201507000-00011",
    "author": Array [
      Object {
        "family": "Peltan",
        "given": "Ithan D.",
      },
      Object {
        "family": "Vande Vusse",
        "given": "Lisa K.",
      },
      Object {
        "family": "Maier",
        "given": "Ronald V.",
      },
      Object {
        "family": "Watkins",
        "given": "Timothy R.",
      },
    ],
    "container-title": "Critical Care Medicine",
    "id": "peltan_international_2015",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "07",
          "",
        ],
      ],
    },
    "number": "7",
    "page": "1429–1438",
    "title": "An International Normalized Ratio–Based Definition of Acute Traumatic Coagulopathy Is Associated With Mortality, Venous Thromboembolism, and Multiple Organ Failure After Injury:",
    "type": "article",
    "volume": "43",
  },
  Object {
    "DOI": "10.1097/TA.0000000000000657",
    "URL": "http://content.wkhealth.com/linkback/openurl?sid=WKPTLP:landingpage&an=01586154-201506000-00023",
    "author": Array [
      Object {
        "family": "Inaba",
        "given": "Kenji",
      },
      Object {
        "family": "Rizoli",
        "given": "Sandro",
      },
      Object {
        "family": "Veigas",
        "given": "Precilla V.",
      },
      Object {
        "family": "Callum",
        "given": "Jeannie",
      },
      Object {
        "family": "Davenport",
        "given": "Ross",
      },
      Object {
        "family": "Hess",
        "given": "John",
      },
      Object {
        "family": "Maegele",
        "given": "Marc",
      },
    ],
    "container-title": "Journal of Trauma and Acute Care Surgery",
    "id": "inaba_2014_2015",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "06",
          "",
        ],
      ],
    },
    "number": "6",
    "page": "1220–1229",
    "title": "2014 Consensus conference on viscoelastic test–based transfusion guidelines for early trauma resuscitation: Report of the panel",
    "type": "article",
    "volume": "78",
  },
]
`;

exports[`BibTeX Parser should parse a valid date regardless of the order of date parts or what is available 1`] = `
Array [
  Object {
    "id": "case_1",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2016",
          "02",
          "",
        ],
      ],
    },
    "type": "article",
  },
  Object {
    "id": "case_2",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "03",
          "",
        ],
      ],
    },
    "type": "book",
  },
  Object {
    "id": "case_3",
    "issued": Object {
      "date-parts": Array [
        Array [
          "",
          "11",
          "",
        ],
      ],
    },
    "type": "report",
  },
  Object {
    "id": "case_4",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2017",
          "",
          "",
        ],
      ],
    },
    "type": "thesis",
  },
  Object {
    "id": "case_5",
    "title": "this case has no listed date",
    "type": "thesis",
  },
]
`;

exports[`BibTeX Parser should parse and skip over preamble statements correctly in varying formats 1`] = `
Array [
  Object {
    "id": "case_1",
    "title": "Test book title",
    "type": "book",
  },
]
`;

exports[`BibTeX Parser should parse dicratical commands 1`] = `
Array [
  Object {
    "id": "simple_dicraticals",
    "title": "fõóōb̀ȧr̂",
    "type": "article",
  },
  Object {
    "id": "extended_dicraticals",
    "title": "föo̧o̱ŏǒọo̊őbąr",
    "type": "article",
  },
]
`;

exports[`BibTeX Parser should parse empty files 1`] = `Array []`;

exports[`BibTeX Parser should parse entries without IDs 1`] = `
Array [
  Object {
    "id": "",
    "title": "Test 1",
    "type": "article",
  },
  Object {
    "id": "",
    "title": "Test 2",
    "type": "book",
  },
  Object {
    "id": "",
    "title": "Test 3",
    "type": "thesis",
  },
  Object {
    "id": "",
    "title": "Test 4",
    "type": "article",
  },
  Object {
    "id": "",
    "title": "Test 5",
    "type": "article",
  },
]
`;

exports[`BibTeX Parser should parse files with CRLF line endings 1`] = `
Array [
  Object {
    "DOI": "10.1097/CCM.0000000000000981",
    "URL": "http://content.wkhealth.com/linkback/openurl?sid=WKPTLP:landingpage&an=00003246-201507000-00011",
    "author": Array [
      Object {
        "family": "Peltan",
        "given": "Ithan D.",
      },
      Object {
        "family": "Vande Vusse",
        "given": "Lisa K.",
      },
      Object {
        "family": "Maier",
        "given": "Ronald V.",
      },
      Object {
        "family": "Watkins",
        "given": "Timothy R.",
      },
    ],
    "container-title": "Critical Care Medicine",
    "id": "peltan_international_2015",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "07",
          "",
        ],
      ],
    },
    "number": "7",
    "page": "1429–1438",
    "title": "An International Normalized Ratio–Based Definition of Acute Traumatic Coagulopathy Is Associated With Mortality, Venous Thromboembolism, and Multiple Organ Failure After Injury:",
    "type": "article",
    "volume": "43",
  },
  Object {
    "DOI": "10.1097/TA.0000000000000657",
    "URL": "http://content.wkhealth.com/linkback/openurl?sid=WKPTLP:landingpage&an=01586154-201506000-00023",
    "author": Array [
      Object {
        "family": "Inaba",
        "given": "Kenji",
      },
      Object {
        "family": "Rizoli",
        "given": "Sandro",
      },
      Object {
        "family": "Veigas",
        "given": "Precilla V.",
      },
      Object {
        "family": "Callum",
        "given": "Jeannie",
      },
      Object {
        "family": "Davenport",
        "given": "Ross",
      },
      Object {
        "family": "Hess",
        "given": "John",
      },
      Object {
        "family": "Maegele",
        "given": "Marc",
      },
    ],
    "container-title": "Journal of Trauma and Acute Care Surgery",
    "id": "inaba_2014_2015",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "06",
          "",
        ],
      ],
    },
    "number": "6",
    "page": "1220–1229",
    "title": "2014 Consensus conference on viscoelastic test–based transfusion guidelines for early trauma resuscitation: Report of the panel",
    "type": "article",
    "volume": "78",
  },
]
`;

exports[`BibTeX Parser should parse files with verbatim tildes, comments, odd characters in keys, super/subscript, no-dot i/j, and math-mode switches 1`] = `
Array [
  Object {
    "URL": "\\\\:",
    "author": Array [
      Object {
        "family": "Giunchiglia",
        "given": "Fausto",
      },
      Object {
        "family": "Traverso",
        "given": "Paolo",
      },
    ],
    "container-title": "Proçeedings of ECP-99",
    "id": "nau-etal:2001_a",
    "issued": Object {
      "date-parts": Array [
        Array [
          "1999",
          "",
          "",
        ],
      ],
    },
    "note": "Paper accompanying invíted talk to be presented at the 1999 European Conference on Planning (ECP-99). Available through http://afrodite.itc.it:1024/~leaf/",
    "publisher": "",
    "title": "Planning as Model-checking",
    "type": "article",
  },
]
`;

exports[`BibTeX Parser should parse files with zero entries 1`] = `Array []`;

exports[`BibTeX Parser should parse jabref files 1`] = `
Array [
  Object {
    "author": Array [
      Object {
        "family": "Doe",
        "given": "John",
      },
      Object {
        "family": "Smith",
        "given": "Jane",
      },
      Object {
        "family": "Dole",
        "given": "Bob",
      },
    ],
    "edition": "1",
    "editor": Array [
      Object {
        "family": "Brown",
        "given": "Pam",
      },
      Object {
        "family": "Waters",
        "given": "Judy",
      },
    ],
    "id": "",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "10",
          "",
        ],
      ],
    },
    "publisher": "Publishing & Co",
    "title": "Testing title of book",
    "type": "book",
    "volume": "2",
  },
]
`;

exports[`BibTeX Parser should parse known commands and skip unknown commands 1`] = `
Array [
  Object {
    "id": "known_commands",
    "note": "© 2017",
    "title": "ΘΣπ",
    "type": "book",
  },
  Object {
    "id": "skip_unknown_commands",
    "title": "Hello World",
    "type": "article",
  },
  Object {
    "id": "symbol_commands",
    "title": "$%#&_{}",
    "type": "article",
  },
]
`;

exports[`BibTeX Parser should parse minified files 1`] = `
Array [
  Object {
    "author": Array [
      Object {
        "family": "Torre",
        "given": "Joe",
      },
      Object {
        "family": "Verducci",
        "given": "Tom",
      },
    ],
    "id": "with_spaces_single",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2008",
          "",
          "",
        ],
      ],
    },
    "publisher": "Doubleday",
    "title": "The Yankee Years",
    "type": "book",
  },
  Object {
    "editor": Array [
      Object {
        "family": "Correns",
        "given": "Paul",
      },
    ],
    "id": "without_spaces_single",
    "page": "1-11",
    "title": "Die dem Boethius fälschlich zugeschriebene Abhandlung des Dominicus Gundisalvi 'De unitate'",
    "type": "article",
    "volume": "1",
  },
  Object {
    "id": "with_spaces_multiple_1",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "",
          "",
        ],
      ],
    },
    "title": "Testing 1",
    "type": "article",
  },
  Object {
    "id": "with_spaces_multiple_2",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "",
          "",
        ],
      ],
    },
    "title": "Testing 2",
    "type": "article",
  },
  Object {
    "id": "without_spaces_multiple_1",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "",
          "",
        ],
      ],
    },
    "title": "Testing 1",
    "type": "book",
  },
  Object {
    "id": "without_spaces_multiple_2",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2015",
          "",
          "",
        ],
      ],
    },
    "title": "Testing 1",
    "type": "book",
  },
]
`;

exports[`BibTeX Parser should parse names in various formats 1`] = `
Array [
  Object {
    "author": Array [
      Object {
        "dropping-particle": "de la Fontaine du",
        "family": "Bois Joli",
        "given": "Jean",
      },
      Object {
        "family": "Diderik van der Waals",
        "given": "Johannes",
      },
      Object {
        "family": "Smith",
        "given": "John",
      },
    ],
    "id": "case_1",
    "type": "article",
  },
  Object {
    "author": Array [
      Object {
        "dropping-particle": "De la Fontaine du",
        "family": "Bois Joli",
        "given": "Jean",
      },
      Object {
        "family": "Waals",
        "given": "Johannes Diderik",
        "non-dropping-particle": "Van der",
      },
      Object {
        "family": "Smith",
        "given": "John",
      },
    ],
    "id": "case_2",
    "type": "article",
  },
  Object {
    "author": Array [
      Object {
        "dropping-particle": "De la Fontaine du",
        "family": "Bois Joli",
        "given": "Jean",
        "suffix": "Jr.",
      },
      Object {
        "family": "Waals",
        "given": "Johannes Diderik",
        "non-dropping-particle": "Van der",
        "suffix": "III",
      },
      Object {
        "family": "Smith",
        "given": "John",
        "suffix": "IV",
      },
    ],
    "id": "case_3",
    "type": "article",
  },
  Object {
    "author": Array [
      Object {
        "family": "Smith",
        "given": "Frank",
      },
      Object {
        "family": "Research Collaborators",
        "given": "Foobar",
      },
    ],
    "id": "case_4",
    "type": "article",
  },
]
`;

exports[`BibTeX Parser should parse non-standard fields 1`] = `
Array [
  Object {
    "DOI": "10.1007/978-981-13-9913-8_2",
    "URL": "https://link.springer.com/chapter/10.1007%2F978-981-13-9913-8_2",
    "author": Array [
      Object {
        "family": "Verkhratsky",
        "given": "Alexei",
      },
      Object {
        "family": "Ho",
        "given": "Margaret S",
      },
      Object {
        "family": "Parpura",
        "given": "Vladimir",
      },
    ],
    "container-title": "Advances in experimental medicine and biology",
    "id": "Verkhratsky2019Evolution",
    "issued": Object {
      "date-parts": Array [
        Array [
          "2019",
          "",
          "",
        ],
      ],
    },
    "page": "15–44",
    "publisher-place": "United States",
    "title": "Evolution of Neuroglia",
    "type": "article",
    "volume": "1175",
  },
]
`;

exports[`BibTeX Parser should parse user-defined strings 1`] = `
Array [
  Object {
    "id": "no_spaces_test",
    "title": "AAABBBCCC",
    "type": "article",
  },
  Object {
    "id": "test_concat_whitespace",
    "title": "foo bar baz",
    "type": "book",
  },
  Object {
    "id": "parses_defined_strings",
    "note": "foo bar baz",
    "title": "foo bar baz",
    "type": "article",
  },
  Object {
    "id": "skips_over_unknown_string",
    "title": "Title starts here",
    "type": "book",
  },
  Object {
    "id": "parses_string_above_now_that_its_available",
    "title": "Now this string is available to use!",
    "type": "article",
  },
]
`;
